File: Easy/159.rs
// Daily Coding Problem: 159
// Difficulty: Easy
// Asked by: Google
// Authored by: TenthEdict

use std::collections::HashSet;

fn first_recurring_char(s: &str) -> Option<char> {
    // 1. Initialize an empty HashSet to track seen characters
    let mut hset = HashSet::new();
    // 2. Iterate through each character in the string
    for c in s.chars() {
        // 3. If we've seen this character before, return it
        if hset.contains(&c) {
            return Some(c);
        } else {
            // 4. Otherwise, add it to our seen set
            hset.insert(c);
        };
    }
    // 5. If we get through the whole string, return None
    return None;
}

fn main() {
    let repeat = "abcbde";
    let norepeat = "abcdef";
    println!("{:?}", first_recurring_char(repeat));
    println!("{:?}", first_recurring_char(norepeat));
}

File: Easy/JavaScript/159.js
/**
 * Daily Coding Problem: 159
 * Difficulty: Easy
 * Asked By: Google
 * Authored By: TenthEdict
 * 
 * Question: Given a string, return the first recurring character in it, or null if there is no recurring character. For example, given the string "acbbac", return "b". Given the string "abcdef", return null.
 * 
 */

function first_recurring_character(str) {
    let seen = {}
    for (i in str) {
        if (str[i] in seen) {
            return str[i]
        } else {
            seen[str[i]] = true
        }
    }
    return null
}

console.log(first_recurring_character("acbbac")) // prints b
console.log(first_recurring_character("abcdef")) // prints null
console.log(first_recurring_character("abcdef")) // prints null
